central/networkgraph/flow/datastore/internal/store/postgres/store.go

```diff
@@ -863,12 +884,27 @@ func (s *flowStoreImpl) pruneEntities(ctx context.Context, deleteStmt string, en
 	ctx, cancel := context.WithTimeout(ctx, deleteTimeout)
 	defer cancel()
 
-	ct, err := conn.Exec(ctx, deleteStmt, entityIds)
+	rows, err := conn.Query(ctx, deleteStmt, entityIds)
 	if err != nil {
 		return 0, err
 	}
 
-	return ct.RowsAffected(), nil
+	deletedIDs, err := s.readIdFromRows(rows)
+	if err != nil {
+		return 0, err
+	}
+
+	// The networktree still has the pruned entities,
+	// so we need to propagate the removal.
+	// TODO: ROX-29450 will make this unnecessary
+	// and we can remove this logic.
+	if networktree := s.networktreeMgr.GetNetworkTree(ctx, s.clusterID.String()); networktree != nil {
+		for _, id := range deletedIDs {
+			networktree.Remove(id)
```

Comment:
I am a little concerned. I think the locks and the way the queries work prevent the database from getting into an inconsistent state. For example if a flow is deleted, such that there is an external entity that it becomes orphaned and then another flow that points to that entity is added in the brief time before the entity is deleted, the entity will not actually be deleted. With the network tree, I don't think there are such protections. What would happen if the network tree had a flow with an external entity, that didn't exist, because it was erroneously deleted?